include/sys/zio.h

```diff
@@ -102,6 +102,30 @@ enum zio_checksum {
 #define	ZIO_DEDUPCHECKSUM	ZIO_CHECKSUM_SHA256
 #define	ZIO_DEDUPDITTO_MIN	100
 
+/* supported encryption algorithms */
+enum zio_encrypt {
+	ZIO_CRYPT_INHERIT = 0,
+	ZIO_CRYPT_ON,
+	ZIO_CRYPT_OFF,
+	ZIO_CRYPT_AES_128_CCM,
+	ZIO_CRYPT_AES_192_CCM,
+	ZIO_CRYPT_AES_256_CCM,
+	ZIO_CRYPT_AES_128_GCM,
+	ZIO_CRYPT_AES_192_GCM,
+	ZIO_CRYPT_AES_256_GCM,
+	ZIO_CRYPT_FUNCTIONS
+};
+
+#define	ZIO_CRYPT_ON_VALUE	ZIO_CRYPT_AES_256_CCM
+#define	ZIO_CRYPT_DEFAULT	ZIO_CRYPT_OFF
+
+/* macros defining encryption lengths */
+#define	MAX_MASTER_KEY_LEN 32
+#define	DATA_IV_LEN 12
+#define	DATA_SALT_LEN 8
+#define	DATA_MAC_LEN 16
```

Comment:
> DATA here (and above) is in contrast to ZIL?

Yes. Unfortunately, there is only room in the `zil_chain_t` for an 8 byte MAC. Ideally we would be able to squeeze the full 16 bytes in there like we do for everything else but I don't see a way to make that happen. GCM and CCM both support an 8 byte MAC (with some reduced security).

> We might want to throw on a ZFS_ or ZIO_ prefix to these macro names.

I like `ZIO_*`. I'll add that in.